Add basic support for white-listing tests to run on new PRs.

samoht opened the pull request:
Currently, the only working option is to say "test this" to allow
a PR to be tested.

Signed-off-by: Thomas Gazagnaire <thomas@gazagnaire.org>

samoht commented:
/cc @dave-tucker The workflow for testing stuff is very basic atm but it should be able to do the basic things needed, e.g. build/test will be done if someone on the whitelist opened the PR or if someone on the whitelist said "test me". The code compiles but is untested ...

Before merging, a `whitelist` file should added at the root of this repo, containing GitHub logins (one per line). DataKit will monitor this repo and will recompute PRs "testability" everytime a new commit is done to this repo.

talex5 commented on src/PR_state.ml:
```diff
@@ -0,0 +1,23 @@
+    let test_this c =
+      (* The user doing the comment is in the whitelist and he/she
+         says "test this" *)
+      in_whitelist (Comment.user c) && contains "test this" (Comment.body c)
```
Looks like Jenkins uses `test this please` - maybe that's less prone to mistakes? (e.g. `Did you test this?`)

talex5 commented on src/linuxkit_build.ml:
```diff
@@ -368,7 +381,7 @@ module Builder = struct
     | Error e -> Utils.failf "Unexpected DB error: %a" DK.pp_error e
 
-  let branch _t { Key.src; target} =
+  let branch _t { Key.src; target; _} =
```
It doesn't make sense to ignore something that's in the key. If you want all PRs, branches and tags to rebuild whenever the whitelist changes, the (hash of) the whitelist needs to be in the branch name.

If you don't, then the whitelist should be in the context.

Most likely, checking whether someone is whitelisted has nothing to do with building LinuxKit and should be checked as a separate (non-cached) build step. Then, changing the whitelist will only rebuild PRs affected by the change.

talex5 commented on src/linuxkit_build.ml:
```diff
@@ -385,14 +398,13 @@ module Result_cache = Cache.Make(Builder)
-  Result_cache.find t job_id key
+  Term.target target >>= fun target_v ->
+  let key = { Builder.Key. src; target; whitelist; } in
+  let context = { Builder.job_id; target = target_v } in
```
I don't think putting the target in the context makes sense either. Whether a PR is built does depend on the state of the PR, not just on the commit.

talex5 reviewed (commented):
I think the whitelist check should be a separate (non-cached) build step. This would avoid duplicating the check in build and test, and should make it rebuild exactly when needed.

samoht committed:
Add .merlin to .gitignore

samoht commented:
As suggested by @talex5, I've added the PR check as a separate pipeline step so it could be cached more easily. The current pipeline becomes:

- pull whitelist from `linuxkit/linuxkit-ci/master/whitelist`
- check that the target (PR, ref) can be built using the information stored in the whitelist
- build and test as it is now

samoht committed:
Add basic support for white-listing tests to run on new PRs.

Currently, the only working option is to say "ok to test" to allow
a PR to be tested.

Signed-off-by: Thomas Gazagnaire <thomas@gazagnaire.org>

dave-tucker commented:
LGTM and a whitelist file now exists in this repo. Is this good to merge/test?

samoht commented:
It is using unreleased features in datakit-github, so I need to add a few pins in the Dockerfile (doing that now). But if you managed to get it compiled it should be good to test.

samoht committed:
Fix Dockerfile

Signed-off-by: Thomas Gazagnaire <thomas@gazagnaire.org>

samoht commented:
I've just fixed the Dockerfile so you should be able to test it more easily.

samoht commented:
What's the current plan regarding LinuxKit CI? Should I close this PR?

justincormack commented:
Hmm, we don't have a very determinate plan over CI at present... so maybe.
